File: 0131-palindrome-partitioning/0131-palindrome-partitioning.cpp
class Solution {
public:
    vector<vector<string>> res;
    vector<string> list;
    vector<vector<string>> partition(string s) {
        backtracking(s, 0);
        return res;
    }

    void backtracking(string s, int start) {
        if (start >= s.size()) {
            res.push_back(list);
            return;
        }

        for (int i = start; i < s.size(); i++) {
            if (isPalindrome(s, start, i)) {
                list.push_back(s.substr(start, i - start + 1));
            } else {
                continue;
            }

            backtracking(s, i + 1);
            list.pop_back();
        }
    }

    bool isPalindrome(string s, int left, int right) {
        while(left < right) {
            if (s[left] == s[right]) {
                left++;
                right--;
            } else {
                return false;
            }
        }

        return true;
    }
};

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        def isPalindrome(left, right):
            while left < right:
                if s[left] != s[right]:
                    return False
                
                left += 1
                right -= 1

            return True

        res = []
        partition = []

        def backtracking(idx):
            if idx == len(s):
                res.append(copy.copy(partition))
                return

            for i in range(idx, len(s)):
                if isPalindrome(idx, i):
                    partition.append(s[idx: i + 1])
                    backtracking(i + 1)
                    partition.pop()

        backtracking(0)

        return res